src/AppInstallerCLICore/Workflows/PinFlow.cpp

```diff
@@ -0,0 +1,247 @@
+            AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_PIN_DOES_NOT_EXIST);
+        }
+
+        pinningIndex->RemovePin(pinKey);
+    }
+
+    // Report the pins in a table.
+    // Required Args: None
+    // Inputs: Pins
+    // Outputs: None
+    void ReportPins(Execution::Context& context)
+    {
+        const auto& pins = context.Get<Execution::Data::Pins>();
+        if (pins.empty())
+        {
+            context.Reporter.Info() << Resource::String::PinNoPinsExist << std::endl;
+            return;
+        }
+
+        // TODO: Use package and source names
+        Execution::TableOutput<4> table(context.Reporter,
+            {
+                Resource::String::SearchId,
+                Resource::String::SearchSource,
+                Resource::String::SearchVersion,
+                Resource::String::PinType,
+            });
+
+        for (const auto& pin : pins)
+        {
+            // TODO: Avoid these conversions to string
+            table.OutputLine({
+                pin.GetPackageId(),
+                std::string{ pin.GetSourceId() },
+                std::string{ pin.GetVersionString() },
+                std::string{ ToString(pin.GetType()) },
+                });
+        }
+
+        table.Complete();
+    }
+
+    // Resets all the existing pins.
+    // Required Args: None
+    // Inputs: PinningIndex
+    // Outputs: None
+    void ResetAllPins(Execution::Context& context)
+    {
+        AICLI_LOG(CLI, Info, << "Resetting all pins");
+        if (context.Args.Contains(Execution::Args::Type::Force))
+        {
+            context.Reporter.Info() << Resource::String::PinResettingAll << std::endl;
+
+            if (context.Get<Execution::Data::Pins>().empty())
+            {
+                context.Reporter.Info() << Resource::String::PinNoPinsExist << std::endl;
+            }
+            else
+            {
+                auto pinningIndex = context.Get<Execution::Data::PinningIndex>();
+                pinningIndex->ResetAllPins();
+            }
+        }
+        else
+        {
+            AICLI_LOG(CLI, Info, << "--force argument is not present");
+            context.Reporter.Info() << Resource::String::PinResetUseForceArg << std::endl;
+            context << ReportPins;
+        }
+    }
+
+    // Updates the list of pins to include only those matching the current open source
+    // Required Args: None
+    // Inputs: Pins, Source
+    // Outputs: None
+    void CrossReferencePinsWithSource(Execution::Context& context)
+    {
+        const auto& pins = context.Get<Execution::Data::Pins>();
+        const auto& source = context.Get<Execution::Data::Source>();
+
+        std::vector<Pinning::Pin> matchingPins;
+        std::copy_if(pins.begin(), pins.end(), std::back_inserter(matchingPins), [&](Pinning::Pin pin) {
+            // TODO: Filter to source
+            SearchRequest searchRequest;

```

Comment:
That seems expensive to do a search for each pin, shall we do this when actually needed during upgrade?
meaning do we need to support listing pins filtered by source?

Reply:
I removed the searches from the case when we are actually resetting the pins (--force). But when we don't have --force, I'm listing the pins and I want to not show pins that are no longer applicable (e.g. because the app was uninstalled or because the source no longer provides the package). I'm assuming there won't be that many pins, but I certainly didn't consider that it would be expensive for REST sources.